experimental/learn_descriptors/hloc_helper.py

```diff
@@ -0,0 +1,254 @@
+
+    def __post_init__(self):
+        assert isinstance(self.dir_images, Path)
+        assert self.dir_images.exists()
+        assert isinstance(self.dir_outputs, Path)
+        assert isinstance(self.path_sfm_pairs, Path)
+        assert isinstance(self.dir_sfm, Path)
+        assert isinstance(self.local_feature_config, dict)
+        assert isinstance(self.matcher_local_config, dict)
+        assert isinstance(self.exhaustive_match, bool)
+        assert self.exhaustive_match or isinstance(self.global_feature_config, dict)
+        assert self.exhaustive_match or isinstance(self.retrieval_num_matched, int)
+        if self.path_local_features is None:
+            self.path_local_features = self.dir_outputs / "features.h5"
+        if self.path_global_features is None:
+            self.path_global_features = self.dir_outputs / "global_features.h5"
+        if self.path_local_matches_sfm is None:
+            self.path_local_matches_sfm = self.dir_outputs / "matches.h5"
+
+
+class HlocHelper:
+    def __init__(
+        self,
+        config: HlocHelperConfig | None = None,
+        dir_cached_colmap_model: pycolmap.Reconstruction | None = None,
+    ):
+        if config is not None:
+            self.config = config
+        if dir_cached_colmap_model is not None:
+            self.model = pycolmap.Reconstruction(str(dir_cached_colmap_model))
+        else:
+            self.model = None
+
+    # images in image_list should be relative to self.config.dir_images
+    def extract_and_match(self, image_list: list[Path]):
+        extract_features.main(
+            self.config.local_feature_config,
+            self.config.dir_images,
+            self.config.dir_outputs,
+            feature_path=self.config.path_local_features,
+            image_list=image_list,
+        )
+        if not self.config.exhaustive_match:
+            extract_features.main(
+                self.config.global_feature_config,
+                self.config.dir_images,
+                self.config.dir_outputs,
+                feature_path=self.config.path_global_features,
+            )
+            pairs_from_retrieval.main(
+                self.config.path_global_features,
+                self.config.path_sfm_pairs,
+                num_matched=self.config.retrieval_num_matched,
+                query_list=image_list,
+                db_list=image_list,
+            )
+        else:
+            pairs_from_exhaustive.main(
+                self.config.path_sfm_pairs,
+                image_list=image_list,
+            )
+        match_features.main(
+            self.config.matcher_local_config,
+            self.config.path_sfm_pairs,
+            self.config.path_local_features,
+            self.config.dir_outputs,
+            self.config.path_local_matches_sfm,
+        )
+
+    # if image_list is left as None, all images will be imported into colmap_db
+    def reconstruct(self, image_list: list[Path] = None) -> pycolmap.Reconstruction:
```

Comment:
add some assert that makes sure this is okay. Maybe check that all of the images have had features extracted?